detection_rules/rule.py

```diff
@@ -700,6 +700,13 @@ class QueryRuleData(BaseRuleData):
     language: definitions.FilterLanguages
     alert_suppression: Optional[AlertSuppressionMapping] = field(metadata=dict(metadata=dict(min_compat="8.8")))
 
+    @cached_property
+    def index_or_dataview(self) -> list[str]:
+        """Return the index or dataview depending on which is set."""
+        if self.index is not None:
+            return self.index
+        return [self.dataview]
```

Comment:
What if `self.dataview` is also None?

Reply:
Then it would fail the validation from https://github.com/elastic/detection-rules/pull/3830.

Reply:
I think it makes sense to default and return `[ ]` from https://github.com/elastic/detection-rules/pull/3857#discussion_r1667084970, which should address this. However, we may also want to consider adding in a validation check to catch this given that I do not think we want to support rules where neither of these are set (no data would pass into it).